### PR TITLE
ARROW-1665: [Serialization] Support more custom datatypes in the default serialization context

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

File: python/pyarrow/serialization.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+_default_serialization_context.register_type(
+     type(lambda: 0), "function",
+     pickle=True)
+
+# ----------------------------------------------------------------------
+# Set up serialization for numpy with dtype object (primitive types are
+# handled efficiently with Arrow's Tensor facilities, see python_to_arrow.cc)
+
+try:
```
Comment (xhochy) [inline python/pyarrow/serialization.py]: `numpy` is a hard requirement of `pyarrow` so there should be no need for the `try…except` here.

**Review by xhochy — CHANGES_REQUESTED** [review]
LGTM except a small nitpick

File: python/pyarrow/serialization.py (hunk 1 above)
Comment (wesm) [reply python/pyarrow/serialization.py]: fixed

**Comment by wesm** [comment]
There seem to be some transient Travis CI problems, restarting the build

**Comment by wesm** [comment]
One of the macOS builds failed for pretty weird and unrelated reasons:

```
dyld: lazy symbol binding failed: Symbol not found: _utimensat
  Referenced from: /usr/local/Cellar/cmake/3.9.4/bin/cmake
  Expected in: /usr/lib/libSystem.B.dylib
dyld: Symbol not found: _utimensat
  Referenced from: /usr/local/Cellar/cmake/3.9.4/bin/cmake
  Expected in: /usr/lib/libSystem.B.dylib
dyld: lazy symbol binding failed: Symbol not found: _utimensat
  Referenced from: /usr/local/Cellar/cmake/3.9.4/bin/cmake
  Expected in: /usr/lib/libSystem.B.dylib
dyld: Symbol not found: _utimensat
  Referenced from: /usr/local/Cellar/cmake/3.9.4/bin/cmake
  Expected in: /usr/lib/libSystem.B.dylib
make[5]: *** [libz.dylib] Abort trap: 6 (core dumped)
make[5]: *** Deleting file `libz.dylib'
make[4]: *** [CMakeFiles/zlib.dir/all] Error 2
make[3]: *** [all] Error 2
make[2]: *** [zlib_ep-prefix/src/zlib_ep-stamp/zlib_ep-build] Error 2
make[1]: *** [CMakeFiles/zlib_ep.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
-- Installing: /Users/travis/build/apache/arrow/cpp-build/googletest_ep-prefix/src/googletest_ep/lib/libgmock_main.a
make[2]: *** [src/arrow/ipc/feather_generated.h] Abort trap: 6 (core dumped)
make[2]: *** Deleting file `src/arrow/ipc/feather_generated.h'
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [src/arrow/ipc/Message_generated.h] Abort trap: 6 (core dumped)
make[2]: *** Deleting file `src/arrow/ipc/Message_generated.h'
make[1]: *** [src/arrow/ipc/CMakeFiles/metadata_fbs.dir/all] Error 2
```

@kou not sure if you have any idea what this is, but I'm going to merge this patch anyway and we can investigate if this error occurs elsewhere

**Pull request closed without merging**